lib/block_view_validator.go

```diff
@@ -0,0 +1,1415 @@
+
+	// UnstakedStakers
+	numUnstakedStakers, err := ReadUvarint(rr)
+	if err != nil {
+		return errors.Wrapf(err, "UnregisterAsValidatorTxindexMetadata.Decode: Problem reading UnstakedStakers: ")
+	}
+	for ii := 0; ii < int(numUnstakedStakers); ii++ {
+		unstakedStaker := &UnstakedStakerTxindexMetadata{}
+		err = unstakedStaker.RawDecodeWithoutMetadata(blockHeight, rr)
+		if err != nil {
+			return errors.Wrapf(err, "UnregisterAsValidatorTxindexMetadata.Decode: Problem reading UnstakedStakers: ")
+		}
+		txindexMetadata.UnstakedStakers = append(txindexMetadata.UnstakedStakers, unstakedStaker)
+	}
+
+	return nil
+}
+
+func (txindexMetadata *UnregisterAsValidatorTxindexMetadata) GetVersionByte(blockHeight uint64) byte {
+	return 0
+}
+
+func (txindexMetadata *UnregisterAsValidatorTxindexMetadata) GetEncoderType() EncoderType {
+	return EncoderTypeUnregisterAsValidatorTxindexMetadata
+}
+
+//
+// DB UTILS
+//
+
+func DBKeyForValidatorByPKID(validatorEntry *ValidatorEntry) []byte {
+	var key []byte
+	key = append(key, Prefixes.PrefixValidatorByPKID...)
+	key = append(key, validatorEntry.ValidatorPKID.ToBytes()...)
+	return key
+}
+
+func DBKeyForValidatorByStake(validatorEntry *ValidatorEntry) []byte {
+	var key []byte
+	key = append(key, Prefixes.PrefixValidatorByStake...)
+	// FIXME: ensure that this left-pads the uint256 to be equal width
+	key = append(key, EncodeUint256(validatorEntry.TotalStakeAmountNanos)...)                 // Highest stake first
+	key = append(key, EncodeUint64(math.MaxUint64-validatorEntry.RegisteredAtBlockHeight)...) // Oldest first
+	key = append(key, validatorEntry.ValidatorPKID.ToBytes()...)
+	return key
+}
+
+func DBKeyForGlobalStakeAmountNanos() []byte {
+	var key []byte
+	key = append(key, Prefixes.PrefixGlobalStakeAmountNanos...)
+	return key
+}
+
+func DBGetValidatorByPKID(handle *badger.DB, snap *Snapshot, pkid *PKID) (*ValidatorEntry, error) {
+	var ret *ValidatorEntry
+	var err error
+	handle.View(func(txn *badger.Txn) error {
+		ret, err = DBGetValidatorByPKIDWithTxn(txn, snap, pkid)
+		return nil
+	})
+	return ret, err
+}
+
+func DBGetValidatorByPKIDWithTxn(txn *badger.Txn, snap *Snapshot, pkid *PKID) (*ValidatorEntry, error) {
+	// Retrieve ValidatorEntry from db.
+	key := DBKeyForValidatorByPKID(&ValidatorEntry{ValidatorPKID: pkid})
+	validatorBytes, err := DBGetWithTxn(txn, snap, key)
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return nil.
+		if err == badger.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetValidatorByPKID: problem retrieving ValidatorEntry")
+	}
+
+	// Decode ValidatorEntry from bytes.
+	validatorEntry := &ValidatorEntry{}
+	rr := bytes.NewReader(validatorBytes)
+	if exist, err := DecodeFromBytes(validatorEntry, rr); !exist || err != nil {
+		return nil, errors.Wrapf(err, "DBGetValidatorByPKID: problem decoding ValidatorEntry")
+	}
+	return validatorEntry, nil
+}
+
+func DBGetTopValidatorsByStake(
```

Comment:
Good point. I will have to punt until a subsequent PR where we implement stake + unstake + unlock stake since currently all validators have zero stake.